app/src/scripts/soc/soc.tsx

```diff
@@ -125,22 +129,46 @@ export abstract class SOC_Generic {
             return [];
 
         const upperPhrase: string = phrase.toUpperCase();
+
+        // Check cache first for both broad and specific course searches
+        // Use course code as cache key for specific searches
+        const cacheKey =
+            searchBy === SearchBy.COURSE_CODE
+                ? upperPhrase
+                : `${searchBy}:${upperPhrase}`;
+        if (this.courseCache.has(cacheKey)) {
+            console.log("Returning cached results for:", cacheKey);
+            // Return cached results
+            return this.courseCache.get(cacheKey) || [];
+        }
+
+        let results: Course[] = [];
         if (searchBy === SearchBy.COURSE_CODE) {
-            return this.courses.filter((c) => c.code.includes(upperPhrase));
+            results = this.courses.filter((c) => c.code.includes(upperPhrase));
+            // Cache each individual course from the broad search result
+            results.forEach((course) =>
+                this.courseCache.set(course.code.toUpperCase(), [course]),
```

Comment:
This can lead to some nasty bugs with distinct courses that share the same course code (e.g. quest courses & special topics).

You cannot assume that a particular course code belongs to only one course.

You should only cache after searching for all possible courses as there may be more than one.